cmd/dataavailability/data_availability_check.go

```diff
@@ -0,0 +1,301 @@
+	if err != nil {
+		panic(err)
+	}
+	dataAvailabilityCheck, err := newDataAvailabilityCheck(ctx, dataAvailabilityCheckConfig)
+	if err != nil {
+		panic(err)
+	}
+	if err := dataAvailabilityCheck.start(ctx); err != nil {
+		panic(err)
+	}
+}
+
+func (d *DataAvailabilityCheck) start(ctx context.Context) error {
+	latestHeader, err := d.l1Client.HeaderByNumber(ctx, big.NewInt(rpc.FinalizedBlockNumber.Int64()))
+	if err != nil {
+		return err
+	}
+	latestBlockNumber := latestHeader.Number.Uint64()
+	oldBlockNumber := latestBlockNumber - 86400 // 12 days old block number
+
+	log.Info("Running new hash data availability check")
+	newHashErr := d.checkDataAvailabilityForNewHashInBlockRange(ctx, latestBlockNumber, oldBlockNumber)
+	log.Info("Completed new hash data availability check")
+
+	log.Info("Running old hash data availability check")
+	oldHashErr := d.checkDataAvailabilityForOldHashInBlockRange(ctx, oldBlockNumber, latestBlockNumber)
+	log.Info("Completed old hash data availability check")
+
+	if newHashErr != nil || oldHashErr != nil {
+		return fmt.Errorf("New Hash Check: %w, Old Hash Check: %s", newHashErr, oldHashErr)
+	}
+	return nil
+}
+
+func (d *DataAvailabilityCheck) checkDataAvailabilityForNewHashInBlockRange(ctx context.Context, latestBlock uint64, oldBlock uint64) error {
+	for latestBlock-d.config.L1BlocksPerRead >= oldBlock {
+		query := ethereum.FilterQuery{
+			FromBlock: new(big.Int).SetUint64(latestBlock - d.config.L1BlocksPerRead),
+			ToBlock:   new(big.Int).SetUint64(latestBlock),
+			Addresses: []common.Address{*d.inboxAddr},
+			Topics:    [][]common.Hash{{batchDeliveredID}},
+		}
+		logs, err := d.l1Client.FilterLogs(ctx, query)
+		if err != nil {
+			return err
+		}
+		for _, deliveredLog := range logs {
+			isDasMessage, err := d.checkDataAvailability(ctx, deliveredLog)
+			if err != nil {
+				return err
+			}
+			if isDasMessage {
+				return nil
+			}
+		}
+		latestBlock = latestBlock - d.config.L1BlocksPerRead
+	}
+	return fmt.Errorf("no das message found between block %d and block %d", latestBlock, oldBlock)
+}
+
+func (d *DataAvailabilityCheck) checkDataAvailabilityForOldHashInBlockRange(ctx context.Context, oldBlock uint64, latestBlock uint64) error {
+	for oldBlock+d.config.L1BlocksPerRead <= latestBlock {
+		query := ethereum.FilterQuery{
+			FromBlock: new(big.Int).SetUint64(oldBlock),
+			ToBlock:   new(big.Int).SetUint64(oldBlock + d.config.L1BlocksPerRead),
+			Addresses: []common.Address{*d.inboxAddr},
+			Topics:    [][]common.Hash{{batchDeliveredID}},
+		}
+		logs, err := d.l1Client.FilterLogs(ctx, query)
+		if err != nil {
+			return err
+		}
+		for _, deliveredLog := range logs {
+			isDasMessage, err := d.checkDataAvailability(ctx, deliveredLog)
+			if err != nil {
+				return err
+			}
+			if isDasMessage {
+				return nil
+			}
+		}
+		oldBlock = oldBlock + d.config.L1BlocksPerRead
+	}
+	return fmt.Errorf("no das message found between block %d and block %d", oldBlock, latestBlock)
```

Comment:
Since `oldBlock` is being modified each iteration, this log message won't show the full range.

Reply:
Thanks for catching this. Fixed now.